google-cloud-spanner/src/main/java/com/google/cloud/spanner/PartitionedDMLTransaction.java

```diff
@@ -60,41 +72,88 @@ private ByteString initTransaction() {
-            Map<String, Value> stmtParameters = statement.getParameters();
-            if (!stmtParameters.isEmpty()) {
-              com.google.protobuf.Struct.Builder paramsBuilder = builder.getParamsBuilder();
-              for (Map.Entry<String, Value> param : stmtParameters.entrySet()) {
-                paramsBuilder.putFields(param.getKey(), param.getValue().toProto());
-                builder.putParamTypes(param.getKey(), param.getValue().getType().toProto());
+    log.log(Level.FINER, "Starting PartitionedUpdate statement");
+    boolean foundStats = false;
+    long updateCount = 0L;
+    Duration remainingTimeout = timeout;
+    Stopwatch stopWatch = Stopwatch.createStarted();
+    try {
+      // Loop to catch AbortedExceptions.
+      while (true) {
+        ByteString resumeToken = ByteString.EMPTY;
+        try {
+          ByteString transactionId = initTransaction();
+          final ExecuteSqlRequest.Builder builder =
+              ExecuteSqlRequest.newBuilder()
+                  .setSql(statement.getSql())
+                  .setQueryMode(QueryMode.NORMAL)
+                  .setSession(session.getName())
+                  .setTransaction(TransactionSelector.newBuilder().setId(transactionId).build());
+          Map<String, Value> stmtParameters = statement.getParameters();
+          if (!stmtParameters.isEmpty()) {
+            com.google.protobuf.Struct.Builder paramsBuilder = builder.getParamsBuilder();
+            for (Map.Entry<String, Value> param : stmtParameters.entrySet()) {
+              paramsBuilder.putFields(param.getKey(), param.getValue().toProto());
+              builder.putParamTypes(param.getKey(), param.getValue().getType().toProto());
+            }
+          }
+          while (true) {
+            remainingTimeout =
+                remainingTimeout.minus(stopWatch.elapsed(TimeUnit.MILLISECONDS), ChronoUnit.MILLIS);
+            try {
+              builder.setResumeToken(resumeToken);
+              ServerStream<PartialResultSet> stream =
+                  rpc.executeStreamingPartitionedDml(
+                      builder.build(), session.getOptions(), remainingTimeout);
+              for (PartialResultSet rs : stream) {
+                if (rs.getResumeToken() != null && !ByteString.EMPTY.equals(rs.getResumeToken())) {
+                  resumeToken = rs.getResumeToken();
+                }
+                if (rs.hasStats()) {
+                  foundStats = true;
+                  updateCount += rs.getStats().getRowCountLowerBound();
+                }
+              }
+              break;
+            } catch (UnavailableException e) {
+              // Retry the stream in the same transaction if the stream breaks with
+              // UnavailableException and we have a resume token. Otherwise, we just retry the
+              // entire transaction.
+              if (resumeToken != null && !ByteString.EMPTY.equals(resumeToken)) {
```

Comment:
nit: I don't know if you need to check for the `resumeToken != null` condition since you always init it to EMPTY and only update it if it's not null.

Reply:
Good point, done.